File: 108.convert-sorted-array-to-binary-search-tree.js
/*
 * @lc app=leetcode id=108 lang=javascript
 *
 * [108] Convert Sorted Array to Binary Search Tree
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {number[]} nums
 * @return {TreeNode}
 */
var nums;
var sortedArrayToBST = function(nums) {
    this.nums = nums;
     return     helper(0,nums.length-1);

};

var helper= function(left,right){
    if(left > right) return null;
    let p = Math.floor((left + right)/2);
    var root = new TreeNode(this.nums[p]);
    root.left = helper(left,p-1);
    root.right = helper(p+1,right);
    return root;
};
// @lc code=end

File: 108.convert-sorted-array-to-binary-search-tree.java
/*
 * @lc app=leetcode id=108 lang=java
 *
 * [108] Convert Sorted Array to Binary Search Tree
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    int nums[];
    public TreeNode helper(int left,int right) {
        if(left>right) return null;
        int mid = (left+right)/2;
        TreeNode root = new TreeNode(nums[mid]);
        root.left = this.helper(left,mid-1);
        root.right = this.helper(mid+1,right);
        return root;
    }
    public TreeNode sortedArrayToBST(int[] nums) {
        this.nums = nums;
        int index = nums.length-1;
        TreeNode root = new TreeNode();

        root =  this.helper(0,index);
        return root;
        
    }
}
// @lc code=end

